desktop/desktopentry/desktopentry.go

```diff
@@ -0,0 +1,218 @@
+	Icon     string
+	Exec     string
+
+	Hidden                bool
+	OnlyShowIn            []string
+	NotShownIn            []string
+	GnomeAutostartEnabled bool
+
+	Actions map[string]*Action
+}
+
+type Action struct {
+	Name string
+	Icon string
+	Exec string
+}
+
+type groupState int
+
+const (
+	unknownGroup groupState = iota
+	desktopEntryGroup
+	desktopActionGroup
+)
+
+func splitStringList(value string) []string {
+	return strings.FieldsFunc(value, func(r rune) bool { return r == ';' })
+}
+
+func Load(filename string) (*DesktopEntry, error) {
+	f, err := os.Open(filename)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	return parse(filename, f)
+}
+
+func parse(filename string, r io.Reader) (*DesktopEntry, error) {
+	de := &DesktopEntry{
+		Filename:              filename,
+		GnomeAutostartEnabled: true,
+	}
+	var (
+		currentGroup          = unknownGroup
+		seenDesktopEntryGroup = false
+		actions               []string
+		currentAction         *Action
+	)
+
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+		// Ignore empty lines and comments
+		if line == "" || strings.HasPrefix(line, "#") {
+			continue
+		}
+		// Start of a new group
+		if strings.HasPrefix(line, "[") {
+			if line == "[Desktop Entry]" {
+				if seenDesktopEntryGroup {
+					return nil, fmt.Errorf("desktop file %q has multiple [Desktop Entry] groups", filename)
+				}
+				seenDesktopEntryGroup = true
+				currentGroup = desktopEntryGroup
+			} else if strings.HasPrefix(line, "[Desktop Action ") && strings.HasSuffix(line, "]") {
+				action := line[len("[Desktop Action ") : len(line)-1]
+				if !strutil.ListContains(actions, action) {
+					return nil, fmt.Errorf("desktop file %q contains unknown action %q", filename, action)
+				}
+				if de.Actions[action] != nil {
+					return nil, fmt.Errorf("desktop file %q has multiple %q groups", filename, line)
+				}
+				currentGroup = desktopActionGroup
+				if de.Actions == nil {
+					de.Actions = make(map[string]*Action, len(actions))
+				}
+				currentAction = &Action{}
+				de.Actions[action] = currentAction
+			} else {
+				// Ignore other groups
+				currentGroup = unknownGroup
+			}
+			continue
+		}
+
+		split := strings.SplitN(line, "=", 2)
+		if len(split) != 2 {
+			return nil, fmt.Errorf("desktop file %q badly formed", filename)
+		}
+		// Trim whitespace around the equals sign
+		key := strings.TrimRight(split[0], "\t\n\v\f\r ")
+		value := strings.TrimLeft(split[1], "\t\n\v\f\r ")
+		switch currentGroup {
+		case desktopEntryGroup:
+			switch key {
+			case "Name":
+				de.Name = value
+			case "Icon":
+				de.Icon = value
+			case "Exec":
+				de.Exec = value
+			case "Hidden":
+				de.Hidden = value == "true"
+			case "OnlyShowIn":
+				de.OnlyShowIn = splitStringList(value)
+			case "NotShownIn":
+				de.NotShownIn = splitStringList(value)
+			case "X-GNOME-Autostart-enabled":
+				de.GnomeAutostartEnabled = value == "true"
```

Comment:
are booleans in this context using only true/false ?

Reply:
The [desktop entry spec](https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#value-types) says:

>  Values of type boolean must either be the string true or false.

However, it also lists this deprecated behaviour:

> Historically some booleans have been represented by the numeric entries 0 or 1. With this version of the standard they are now to be represented as a boolean string. However, if an implementation is reading a pre-1.0 desktop entry, it should interpret 0 and 1 as false and true, respectively. 

I adapted this bit of code from the autostart desktop entry parser, which didn't cover the deprecated syntax. The glib implementation covers 0/1, but I'm not sure how common it is in practice:

https://gitlab.gnome.org/GNOME/glib/-/blob/main/glib/gkeyfile.c#L4598

This is syntax that has been deprecated for more than two decades, and I don't think I've seen it in use the wild. The [autostart spec](https://specifications.freedesktop.org/autostart-spec/autostart-spec-latest.html#idm44929561097088) also mentions the `true` value explicitly, but I'm unsure whether it is trying to add further constraints or just assuming non-deprecated behaviour.